google/resource_google_project_default_service_accounts.go

```diff
@@ -0,0 +1,222 @@
+	}
+
+	return nil
+}
+
+func resourceGoogleProjectDefaultServiceAccountsCreate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+	userAgent, err := generateUserAgentString(d, config.userAgent)
+	if err != nil {
+		return err
+	}
+	pid, ok := d.Get("project").(string)
+	if !ok {
+		return fmt.Errorf("cannot get project")
+	}
+	action, ok := d.Get("action").(string)
+	if !ok {
+		return fmt.Errorf("cannot get action")
+	}
+
+	serviceAccounts, err := resourceGoogleProjectDefaultServiceAccountsList(config, d, userAgent)
+	if err != nil {
+		return fmt.Errorf("error listing service accounts on project %s: %v", pid, err)
+	}
+	changedServiceAccounts := make(map[string]interface{})
+	for _, sa := range serviceAccounts {
+		// As per documentation https://cloud.google.com/iam/docs/service-accounts#default
+		// we have just two default SAs and the e-mail may change. So, it is been filtered
+		// by the Display Name
+		switch strings.ToLower(sa.DisplayName) {
+		case "compute engine default service account":
+			changedServiceAccounts[sa.UniqueId] = sa.Email
+			err := resourceGoogleProjectDefaultServiceAccountsDoAction(d, meta, action, sa.UniqueId, sa.Email, pid)
+			if err != nil {
+				return fmt.Errorf("error doing action %s on Service Account %s: %v", action, sa.Email, err)
+			}
+		case "app engine default service account":
+			changedServiceAccounts[sa.UniqueId] = sa.Email
+			err := resourceGoogleProjectDefaultServiceAccountsDoAction(d, meta, action, sa.UniqueId, sa.Email, pid)
+			if err != nil {
+				return fmt.Errorf("error doing action %s on Service Account %s: %v", action, sa.Email, err)
+			}
+		default:
+			continue
+		}
+	}
+	if err := d.Set("service_accounts", changedServiceAccounts); err != nil {
+		return fmt.Errorf("error setting service_accounts: %s", err)
+	}
+	d.SetId(prefixedProject(pid))
+
+	return nil
+}
+
+func resourceGoogleProjectDefaultServiceAccountsList(config *Config, d *schema.ResourceData, userAgent string) ([]*iam.ServiceAccount, error) {
+	pid, ok := d.Get("project").(string)
+	if !ok {
+		return nil, fmt.Errorf("cannot get project")
+	}
+	response, err := config.NewIamClient(userAgent).Projects.ServiceAccounts.List(prefixedProject(pid)).Do()
+	if err != nil {
+		return nil, fmt.Errorf("failed to list service accounts on project %q: %v", pid, err)
+	}
+	return response.Accounts, nil
+}
+
+func resourceGoogleProjectDefaultServiceAccountsDelete(d *schema.ResourceData, meta interface{}) error {
+	if d.Get("restore_policy").(string) == "NONE" {
+		d.SetId("")
+		return nil
+	}
+
+	pid, ok := d.Get("project").(string)
+	if !ok {
+		return fmt.Errorf("cannot get project")
+	}
+	for saUniqueID, saEmail := range d.Get("service_accounts").(map[string]interface{}) {
+		action := d.Get("action").(string)
+		switch action {
+		case "DISABLE":
+			action := "ENABLE"
+			err := resourceGoogleProjectDefaultServiceAccountsDoAction(d, meta, action, saUniqueID, saEmail.(string), pid)
```

Comment:
This is also a situation where the code inside the case statements are nearly identical. A simpler version would be:
```
origAction := d.Get("action").(string)
newAction := ""
if origAction == "DISABLE" {
  newAction = "ENABLE"
} else if origAction == "DELETE" {
  newAction = "UNDELETE"
} else {
  // error
}
err := resourceGoogleProjectDefaultServiceAccountsDoAction(d, meta, newAction, saUniqueID, saEmail.(string), pid)
```

Reply:
thank you! updated. we should not error in this case, that's the only difference.